File: 1865-finding-pairs-with-a-certain-sum/main.py
class FindSumPairs:

    def __init__(self, nums1: List[int], nums2: List[int]):
        self.nums1 = nums1
        self.nums2 = nums2    

        self.counter2 = Counter(nums2)    

    def add(self, index: int, val: int) -> None:
        org = self.nums2[index]
        self.counter2[org] -= 1
        if self.counter2[org] == 0:
            del self.counter2[org]
        
        self.nums2[index] += val
        self.counter2[self.nums2[index]] += 1

    def count(self, tot: int) -> int:
        return sum(
            self.counter2.get(tot - num, 0) for num in self.nums1
        )


# Your FindSumPairs object will be instantiated and called as such:
# obj = FindSumPairs(nums1, nums2)
# obj.add(index,val)
# param_2 = obj.count(tot)

File: 1865-finding-pairs-with-a-certain-sum/main.go
type FindSumPairs struct {
    nums1 []int
    nums2 []int

    counter2 map[int]int
}

func Constructor(nums1 []int, nums2 []int) FindSumPairs {

    counter := make(map[int]int)
	for _, num := range nums2 {
		counter[num]++
	}

	return FindSumPairs{
		nums1:    nums1,
		nums2:    nums2,
		counter2: counter,
	}
}

func (this *FindSumPairs) Add(index int, val int)  {
    original := this.nums2[index]
	this.counter2[original]--
	if this.counter2[original] == 0 {
		delete(this.counter2, original)
	}

	this.nums2[index] += val
	this.counter2[this.nums2[index]]++
}

func (this *FindSumPairs) Count(tot int) int {
    count := 0
	for _, num := range this.nums1 {
		need := tot - num
		if c, exists := this.counter2[need]; exists {
			count += c
		}
	}
	return count
}


/**
 * Your FindSumPairs object will be instantiated and called as such:
 * obj := Constructor(nums1, nums2);
 * obj.Add(index,val);
 * param_2 := obj.Count(tot);
 */
